### ISSUE TITLE
update_screening.py

### ISSUE THREAD

**Issue by ftiger70** [ftiger70/Euro-Screening#1]
import sys
import datetime as dt
import numpy as np
import pandas as pd
import yfinance as yf
import openpyxl

DEFAULT_FILE = "Put_Credit_Spread_Eurostoxx50_Tool.xlsx"
SHEET_NAME = "2. Screening Settimanale"
TICKER = "^STOXX50E"          # Eurostoxx 50 su Yahoo Finance
DISPLAY_NAME = "EUROSTOXX50 (SX5E)"
FIRST_DATA_ROW = 6            # prima riga dati nel foglio (la riga 6 e' l'esempio)
LAST_DATA_ROW = 60            # deve combaciare con il template

def compute_indicators(df: pd.DataFrame) -> dict:
    """df deve avere colonne Open, High, Low, Close (indice = date, ordine crescente)."""
    close = df["Close"]
    high = df["High"]
    low = df["Low"]

    # RSI (14) - smoothing di Wilder
    delta = close.diff()
    gain = delta.clip(lower=0)
    loss = -delta.clip(upper=0)
    avg_gain = gain.ewm(alpha=1 / 14, min_periods=14, adjust=False).mean()
    avg_loss = loss.ewm(alpha=1 / 14, min_periods=14, adjust=False).mean()
    rs = avg_gain / avg_loss
    rsi = 100 - (100 / (1 + rs))

    # Volatilita' storica 30gg annualizzata (da rendimenti log)
    log_ret = np.log(close / close.shift(1))
    hv30 = log_ret.rolling(30).std() * np.sqrt(252)

    # ATR (14) - Wilder
    prev_close = close.shift(1)
    tr = pd.concat([
        (high - low),
        (high - prev_close).abs(),
        (low - prev_close).abs(),
    ], axis=1).max(axis=1)
    atr14 = tr.ewm(alpha=1 / 14, adjust=False).mean()

    # Medie mobili
    ma50 = close.rolling(50).mean()
    ma200 = close.rolling(200).mean()

    # Range 52 settimane (circa 252 sedute)
    min52 = low.rolling(252, min_periods=50).min()
    max52 = high.rolling(252, min_periods=50).max()

    # Supporto / Resistenza 60gg
    supporto60 = low.rolling(60).min()
    resistenza60 = high.rolling(60).max()

    return dict(
        prezzo=close.iloc[-1],
        rsi=rsi.iloc[-1],
        hv30=hv30.iloc[-1],
        atr14=atr14.iloc[-1],
        ma50=ma50.iloc[-1],
        ma200=ma200.iloc[-1],
        min52=min52.iloc[-1],
        max52=max52.iloc[-1],
        supporto60=supporto60.iloc[-1],
        resistenza60=resistenza60.iloc[-1],
    )

def find_next_empty_row(ws) -> int:
    for r in range(FIRST_DATA_ROW, LAST_DATA_ROW + 1):
        # colonna C = Prezzo Attuale: se vuota consideriamo la riga libera
        # (la riga 6 di esempio ha gia' un valore quindi viene saltata)
        if ws.cell(row=r, column=3).value in (None, ""):
            return r
    return -1

def main():
    filepath = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_FILE

    print(f"Scarico dati storici per {TICKER} ...")
    data = yf.download(TICKER, period="2y", interval="1d", progress=False, auto_adjust=False)
    if data.empty:
        print("ERRORE: nessun dato scaricato. Controlla la connessione o il ticker.")
        sys.exit(1)

    # yfinance a volte restituisce colonne MultiIndex se si passano piu' ticker: normalizza
    if isinstance(data.columns, pd.MultiIndex):
        data.columns = data.columns.get_level_values(0)

    ind = compute_indicators(data)
    for k, v in ind.items():
        if pd.isna(v):
            print(f"ERRORE: dati storici insufficienti per calcolare '{k}'. "
                  f"Serve piu' storico (almeno 252 sedute per MA200/52w).")
            sys.exit(1)

    print("Indicatori calcolati:")
    for k, v in ind.items():
        print(f"  {k}: {v:.4f}")

    print(f"\nApro il file: {filepath}")
    wb = openpyxl.load_workbook(filepath)
    if SHEET_NAME not in wb.sheetnames:
        print(f"ERRORE: foglio '{SHEET_NAME}' non trovato nel file.")
        sys.exit(1)
    ws = wb[SHEET_NAME]

    row = find_next_empty_row(ws)
    if row == -1:
        print("ATTENZIONE: nessuna riga libera nel template. "
              "Aggiungi manualmente altre righe con le stesse formule delle precedenti "
              "(seleziona l'ultima riga compilata e trascina in basso), poi rilancia lo script.")
        sys.exit(1)

    today = dt.date.today()
    ws.cell(row=row, column=1, value=DISPLAY_NAME)                    # A Sottostante
    ws.cell(row=row, column=2, value=today).number_format = "yyyy-mm-dd"  # B Data
    ws.cell(row=row, column=3, value=round(float(ind["prezzo"]), 2))  # C Prezzo
    ws.cell(row=row, column=4, value=round(float(ind["rsi"]), 1))     # D RSI14
    ws.cell(row=row, column=6, value=round(float(ind["hv30"]), 4))    # F Volatilita storica
    ws.cell(row=row, column=8, value=round(float(ind["atr14"]), 2))   # H ATR14
    ws.cell(row=row, column=10, value=round(float(ind["ma50"]), 2))   # J MA50
    ws.cell(row=row, column=11, value=round(float(ind["ma200"]), 2))  # K MA200
    ws.cell(row=row, column=14, value=round(float(ind["min52"]), 2))  # N Min 52 sett.
    ws.cell(row=row, column=15, value=round(float(ind["max52"]), 2))  # O Max 52 sett.
    ws.cell(row=row, column=18, value=round(float(ind["supporto60"]), 2))    # R Supporto 60gg
    ws.cell(row=row, column=19, value=round(float(ind["resistenza60"]), 2))  # S Resistenza 60gg
    ws.cell(row=row, column=26, value="Auto-generato da script (verifica sempre su MEXEM prima di operare)")  # Z Note

    wb.save(filepath)
    print(f"\nFatto! Riga {row} aggiornata in '{SHEET_NAME}'.")
    print("Apri il file: le colonne di punteggio/valutazione/strike consigliati "
          "si aggiorneranno da sole (formule gia' presenti).")

if __name__ == "__main__":
    main()
